broadinstitute/single_cell_portal_core
# Clientside throttling for Sentry error logging (SCP-4335)

This adds frontend sampling controls to Sentry logs, improving signal-to-noise ratio for error observability.

Sentry error logging from the browser was introduced in #1482.  This refines that by throttling certain high-volume events:

* TypeError: Failed to fetch: Error in JavaScript when logging event to Bard
* TypeError: NetworkError when attempting to fetch resource.: Error in JavaScript when logging event to Bard

Only 5% of those will now be logged.  In the past 14 days while logging nearly 100% of them, we've accrued over 6000 such events.  That's too much of our nominal quota.  The new 0.05 sample rate means we expect roughly 300 in the next two weeks.  That lower-but-meaningful number helps mitigate Sentry quota exhaustion while not suppressing all such events.  Full suppression might make sense once we further investigate causes and ticket work.

Other noisy Sentry logs are now fully suppressed:
* Sentry errors and traces in `development` and `test` environments (only some such events were suppressed before)
* The high-volume, low-impact error described in SCP-4360

Like Ruby Sentry logging, JavaScript Sentry logging also now prints to the console when suppressed.  Automated tests verify (and document by example) complex new logic in `shouldLog`.

To manually test:
1. Open DevTools in browser
2. Load home page
4. Open `StudyResults.jsx` and change line 77 to `<StudyComponentIdonotExist`
5. Confirm warning message in console "Suppressing error report to Sentry: in an unlogged environment"
6. Go to Network tab
7. Filter to "sentry"
8. Confirm only 1 request is sent to Sentry, with POST body including `{"type":"client_report"}`
9. Edit line 63 in `sentry-logging.js` to return `false`
10. In reloaded page, confirm multiple requests are sent to Sentry
11. Undo steps 4 and 9

This satisfies SCP-4335.
